From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Nelo-T. Wallus" <n.wallus@sap.com>
Date: Thu, 19 Mar 2026 11:17:37 +0100
Subject: [PATCH] Enable GlobalServiceAccount

Signed-off-by: Nelo-T. Wallus <red.brush9525@fastmail.com>
Signed-off-by: Nelo-T. Wallus <n.wallus@sap.com>
---
 pkg/features/kube_features.go | 1 +
 1 file changed, 1 insertion(+)

diff --git a/pkg/features/kube_features.go b/pkg/features/kube_features.go
--- a/pkg/features/kube_features.go
+++ b/pkg/features/kube_features.go
@@ -2384,6 +2384,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 	GlobalServiceAccount: {
 		{Version: version.MustParse("1.31"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.35"), Default: true, PreRelease: featuregate.Beta},
 	},
 }
 
