From d3ceb941f96b7b16d65d532cee1c0cbb0954b517 Mon Sep 17 00:00:00 2001
From: "Nelo-T. Wallus" <n.wallus@sap.com>
Date: Thu, 19 Mar 2026 11:17:37 +0100
Subject: [PATCH 62/73] Enable GlobalServiceAccount

Signed-off-by: Nelo-T. Wallus <red.brush9525@fastmail.com>
Signed-off-by: Nelo-T. Wallus <n.wallus@sap.com>
---
 pkg/features/kube_features.go | 1 +
 1 file changed, 1 insertion(+)

diff --git a/pkg/features/kube_features.go b/pkg/features/kube_features.go
index 5ebef422aa0..61076671a67 100644
--- a/pkg/features/kube_features.go
+++ b/pkg/features/kube_features.go
@@ -2164,6 +2164,7 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 	GlobalServiceAccount: {
 		{Version: version.MustParse("1.31"), Default: false, PreRelease: featuregate.Alpha},
+		{Version: version.MustParse("1.35"), Default: true, PreRelease: featuregate.Beta},
 	},
 }
 
-- 
2.54.0

