From d1884d86b7271b7eaafee1085cc02515e7b1250a Mon Sep 17 00:00:00 2001
From: "Nelo-T. Wallus" <n.wallus@sap.com>
Date: Wed, 6 May 2026 17:03:58 +0200
Subject: [PATCH 90/90] fixup de19ab00b0d

Signed-off-by: Nelo-T. Wallus <red.brush9525@fastmail.com>
Signed-off-by: Nelo-T. Wallus <n.wallus@sap.com>
---
 pkg/controlplane/apiserver/server.go | 1 -
 1 file changed, 1 deletion(-)

diff --git a/pkg/controlplane/apiserver/server.go b/pkg/controlplane/apiserver/server.go
index 7ee1003b6c4..c3b12017c13 100644
--- a/pkg/controlplane/apiserver/server.go
+++ b/pkg/controlplane/apiserver/server.go
@@ -38,7 +38,6 @@ import (
 	serverstorage "k8s.io/apiserver/pkg/server/storage"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	clientgoinformers "k8s.io/client-go/informers"
-	"k8s.io/client-go/kubernetes"
 	"k8s.io/component-helpers/apimachinery/lease"
 	"k8s.io/klog/v2"
 	"k8s.io/utils/clock"
-- 
2.54.0

